src/Lean/Meta/InferType.lean

```diff
@@ -241,9 +249,9 @@ partial def isPropQuick : Expr → MetaM LBool
   | .proj ..          => return LBool.undef
   | .forallE _ _ b _  => isPropQuick b
   | .mdata _ e        => isPropQuick e
-  | .const c lvls     => do let constType ← inferConstType c lvls; isArrowProp constType 0
```

Comment:
The level instantiation here shouldn't be so bad since it's cached.

Reply:
Do you mean there is a qualitative difference to the one in `isProofQuickApp`?
In general the cache seems extremely effective and efficient (otherwise we would be completely screwed), but there are still misses from backtracking and fresh level mvars.

Reply:
Yes, the one in `isProofQuickApp` applies instantiation to a random universe level term, which is not cached at all.

But the one here goes through the instantiation cache for constant infos, and should almost always be a hit since we've probably inferred the type of the constant before.

Reply:
I feel like I'm missing something :smile: . But since we often have constants of the form `c.{us} : ... -> C.{us}`, it seems to me that both now-eliminated `inferConstType` calls were in danger of cache misses on level mvars.

Reply:
I don't think we can avoid the instantiation here.  Even if we avoid it in `isProp`, we usually do `inferType` to unify the types during unification and there we certainly need to instantiate.  (That is, we can choose between a cache miss here or elsewhere but the total number of cache misses won't go down.)

Reply:
Alright, that makes sense